### ISSUE TITLE
Problem determining width for uncommon types having element of generic T

### ISSUE THREAD

**Issue by sammy-hughes** [golang/go#48311]
### What version of Go are you using (`go version`)?

go version go1.17 , passing compiler option "-gcflags=-G=3"

<pre>
$ go version 
go version go1.17 
$ go run -gcflags=-G=3 main.go
</pre>

### Does this issue reproduce with the latest release?

It is an issue with the soft-released generics feature.

### What operating system and processor architecture are you using (`go env`)?

darwin/amd64

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/eixvc/Library/Caches/go-build"
GOENV="/Users/eixvc/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/eixvc/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/eixvc/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/Cellar/go/1.17/libexec"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/Cellar/go/1.17/libexec/pkg/tool/darwin_amd64"
GOVCS=""
GOVERSION="go1.17"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/Me/projects/test2/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch x86_64 -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/24/k0s47ybx6rv_jgb7x0jnbr8c8k834h/T/go-build4055043970=/tmp/go-build -gno-record-gcc-switches -fno-common"
</pre></details>

### What did you do?
The below snippet doesn't compile in the playground, and does require the "-gcflags=-G=3" flag at compile. I've seen folk configure the playground for Go2/experimental features, but I'm not familiar.

Basically, I wanted an iterator chain. Kinda dirty, yeah, but there were several issues. The biggest issue for me was allocation of generic-bound function on a struct, as element-type of a slice. The type-displacement of an eface wrapper and the un-typing of unsafe.Pointer both served to confirm the scope of issue. I also had some trouble with, I suppose, symbol resolution related to the stenciling of the generic-bound call. 

I suspected the anonymous struct definition may have been the cause, and I did up a (likely more suitable) version using a statically defined type have a slice of eface instead of slice of func()Y, and I still observed issues with indexing and extracting the generic function from eface. I could include that too, but it essentially only served to prove the anonymous struct not at fault.

Needs compiler flags to build. 
https://play.golang.org/p/dhH8Fu6oTOX

It's worth noting that I'm Effing Stoked (TM) that generics are pseudo-supported in Go, as of 1.17. I'm itching for official support in 1.18, and I've been so since like, I don't know, June of 2020? It's so good to see this finally hitting the streets.

### What did you expect to see?

I expected to be able to treat compound values involving generic types in a manner indistinguishable from static types, while the generic types were in scope.

### What did you see instead?
I observed several issues:
1. Observed issues with allocation, due to the generic type information being dropped. `internal compiler error: width not calculated: struct { At int; Ok []func() bool; Do []*func() bool }`
2. Could not call the eface-wrapped generic-bound func()Y directly on type-asserted value `internal compiler error: Unexpected op with CALL during stenciling: DOTTYPE`
3. Could not invoke the generic-bound func()Y as an element of slice of func()Y. `internal compiler error: Unexpected op with CALL during stenciling: INDEX`

**Comment by seankhliao** [comment]
Does this reproduce on tip?
-G=3 in 1.17 is not supported

**Comment by robpike** [comment]
This is what I see at (well, near) tip:

```
# command-line-arguments
./z.go:90:112: undefined: Y
./z.go:97:10: not enough arguments in call to strategy_gate
./z.go:106:9: undefined: moped
./z.go:107:9: undefined: moped
./z.go:108:9: undefined: moped
./z.go:110:9: not enough arguments in call to strategy_gate
```

The error on line 90 does indeed seem problematic.

However, this example is inconveniently large. Can you please cut it down to a smaller program that isolates the issue better?

**Comment by ianlancetaylor** [comment]
I don't understand the test case.  I guess it has been edited?  The current test case doesn't compile at all.

```
foo.go:11:8: undefined: unsafe
foo.go:15:31: undefined: unsafe
foo.go:26:23: undefined: unsafe
foo.go:51:19: undefined: newIterChain
foo.go:52:9: undefined: moped
```

**Comment by sammy-hughes** [comment]
Apologies I added a better example. I'd added a selector in the snippet to switch between broken and working examples, but I didn't actually run, it and did not observe that I also included a reference to sample slice-iterator. I've updated it further, @ianlancetaylor, @robpike with some explicit comments documenting my intended API and the specific code which caused the issues in my example.

**Comment by ianlancetaylor** [comment]
Thanks.  The updated example does compile with tip, so I'm no longer sure that this issue is about.

**Comment by sammy-hughes** [comment]
Hence the comments.

1. Slice of element func()Y as property of anonymous struct causes a sizing issue.
2. Invoking a specific index of a slice of element func()Y causes stenciling error.
3. Invoking a specific index of an interface with underlying type func()Y, type asserted inline, causes a stenciling error.
4. Freshly observed: invoking the inline dereferenced cast of an unsafe pointer cast as pointer to func()Y causes a stenciling error.

**Comment by sammy-hughes** [comment]
Stenciling and sizing issues have apparently been addressed post 1.17 release. One more thing to be excited about, coming in 1.18

**Comment by sammy-hughes** [comment]
Stenciling and sizing issues have apparently been addressed post 1.17 release. One more thing to be excited about, coming in 1.18
